streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -1720,6 +1720,71 @@ object ZStream extends ZStreamPlatformSpecific {
         }
     }
 
+  /**
+   * Combines two streams `s1` and `s2` deterministically using the stream of
+   * boolean values `b` to control which stream to pull from next. `true`
+   * indicates to pull from the left stream and `right` indicates to pull from
+   * the right stream. Only consumes as many elements as requested by `b`. If
+   * either `s1` or `s2` are exhausted further requests for values from them
+   * will be ignored.
+   */
+  final def interleave[R, E, A](
+    b: ZStream[R, E, Boolean],
```

Comment:
Is this expressive enough to handle two streams with different lengths? if all the caller of this combinator can do is provide a stream of booleans, surely they can't know if one of the streams ended and therefore they will not be able to "switch over" to reading just one side.

Maybe we need this to be a function that uses the queues themselves?

Reply:
I was thinking that if the stream of booleans eventually requested values from both side like `Stream(true, false).forever` then the caller would be able to get all the values, but it is a little indirect.

What if we added a parameter `f: (ZStream[R, E, A], ZStream[R, E, A]) => ZStream[R, E, A)` where the first stream is the result of interleaving and the second is the "leftover" stream? So if the caller wanted to pull all remaining elements from the leftover stream they could use `_ ++ _`, if they wants to ignore to ignore them they could use `_._1`, or they could apply more complex logic.

Reply:
Sorry, I read more carefully and saw that you skip over a side that has ended. That makes sense! I think this is fine in this form then.

I suggest that we have a shorthand combinator for the common case of `Stream(false, true).forever` though :-) This combinator can be called `interleaveWith` and the shorter one `interleave`.

Reply:
I agree.